### DISCUSSION TITLE
support of "all" function

### DISCUSSION THREAD

**Discussion by Groni3000** [microsoft/pyright#5867] [Q&A] [answered]
Why pyright doesn't support "all" function with typeguards? Is it too hard to implement? Or it is intended?
For example:
```python
def f(x:float|pd.Series, y:float|pd.Series):
    if all(isinstance(el, pd.Series) for el in [x, y]):
        print(x.shape, y.shape) #Error Cannot access member "shape" for type float    Member "shape" is unknown
    if isinstance(x, pd.Series) and isinstance(y, pd.Series):
        print(x.shape, y.shape) #Works fine cuz "isinstance" is supported type guard
```

And how to be in cases with a lot of elements? Don't really want to check everything by hand.

**Comment by erictraut** [comment, marked as answer]
Type guards are specific to the syntax form used, so it wouldn't be enough to simply implement "support for `all`). In your example above, support would need to be added specifically for `all(isinstance(a, b) for a in [x, y])`. This specific expression form is rarely used, so it wouldn't make sense to add the custom logic to support it.

Supported type guard patterns are much simpler than this. For a full list of supported type guard patterns, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards).

For more complex type guards, pyright provides support for PEP 647 (user-defined type guards) which allow you to define your own type guard logic.

**Reply by Groni3000** [reply]
Thank you for the answer! PEP 647 user-defined type guards works perfectly.
